common/sgx/eeid.c

```diff
@@ -0,0 +1,499 @@
+            SGX_SECINFO_TCS,
+            true));
+
+        vaddr += OE_PAGE_SIZE;
+
+        for (size_t i = 0; i < 2; i++)
+            ADD_PAGE(blank_pg, true);
+        vaddr += OE_PAGE_SIZE; // guard
+        for (size_t i = 0; i < 2; i++)
+            ADD_PAGE(blank_pg, true);
+    }
+
+    oe_sha256_final(&hctx, computed_enclave_hash);
+
+done:
+    return OE_OK;
+}
+
+static bool is_zero(const uint8_t* buf, size_t sz)
+{
+    while (sz != 0)
+        if (buf[--sz] != 0)
+            return false;
+    return true;
+}
+
+static size_t _serialized_eeid_size(const oe_eeid_t* eeid)
+{
+    return 2 * (sizeof(oe_eeid_t) + eeid->data_size) + 8 + 1;
+}
+
+static oe_result_t verify_base_image_signature(const sgx_sigstruct_t* sigstruct)
+{
+    oe_result_t result = OE_UNEXPECTED;
+    unsigned char buf[sizeof(sgx_sigstruct_t)];
+    size_t n = 0;
+
+    OE_CHECK(oe_memcpy_s(
+        buf,
+        sizeof(buf),
+        sgx_sigstruct_header(sigstruct),
+        sgx_sigstruct_header_size()));
+    n += sgx_sigstruct_header_size();
+    OE_CHECK(oe_memcpy_s(
+        &buf[n],
+        sizeof(buf) - n,
+        sgx_sigstruct_body(sigstruct),
+        sgx_sigstruct_body_size()));
+    n += sgx_sigstruct_body_size();
+
+    OE_SHA256 msg_hsh;
+    oe_sha256_context_t context;
+
+    oe_sha256_init(&context);
+    oe_sha256_update(&context, buf, n);
+    oe_sha256_final(&context, &msg_hsh);
+
+    uint8_t reversed_modulus[OE_KEY_SIZE];
+    for (size_t i = 0; i < OE_KEY_SIZE; i++)
+        reversed_modulus[i] = sigstruct->modulus[OE_KEY_SIZE - 1 - i];
+
+    uint8_t reversed_exponent[OE_KEY_SIZE];
+    for (size_t i = 0; i < OE_EXPONENT_SIZE; i++)
+        reversed_exponent[i] = sigstruct->exponent[OE_EXPONENT_SIZE - 1 - i];
+
+    uint8_t reversed_signature[OE_KEY_SIZE];
+    for (size_t i = 0; i < OE_KEY_SIZE; i++)
+        reversed_signature[i] = sigstruct->signature[OE_KEY_SIZE - 1 - i];
+
+    oe_rsa_public_key_t pk;
+#ifdef OE_BUILD_ENCLAVE
+    mbedtls_pk_context pkctx;
+    mbedtls_pk_init(&pkctx);
+    const mbedtls_pk_info_t* info = mbedtls_pk_info_from_type(MBEDTLS_PK_RSA);
+    mbedtls_pk_setup(&pkctx, info);
+
+    mbedtls_rsa_context* rsa_ctx = mbedtls_pk_rsa(pkctx);
+    mbedtls_rsa_init(rsa_ctx, 0, 0);
+    mbedtls_rsa_import_raw(
+        rsa_ctx,
+        reversed_modulus,
+        OE_KEY_SIZE, // N
+        NULL,
+        0,
+        NULL,
+        0,
+        NULL,
+        0, // P Q D
+        reversed_exponent,
+        OE_EXPONENT_SIZE);
+    if (mbedtls_rsa_check_pubkey(rsa_ctx) != 0)
+        OE_RAISE(OE_INVALID_PARAMETER);
+    mbedtls_pk_context* ikey = &pkctx;
+#else
+#if OPENSSL_VERSION_NUMBER < 0x1010100fL
+#error OpenSSL 1.0.2 not supported
+#endif
+    BIGNUM* rm = BN_bin2bn(reversed_modulus, OE_KEY_SIZE, 0);
+    BIGNUM* re = BN_bin2bn(reversed_exponent, OE_EXPONENT_SIZE, 0);
+    RSA* rsa = RSA_new();
+    RSA_set0_key(rsa, rm, re, NULL);
+    EVP_PKEY* ikey = EVP_PKEY_new();
+    EVP_PKEY_assign_RSA(ikey, rsa);
```

Comment:
Hmm, it does seem like there is a small gap in the oe crypto code. There doesn't seem to be an API to create an oe public key from a sigstruct.

For now I'm ok if EEID is linux-only. But the build shouldn't fail on Windows and any documentation should call out that fact.

Adding openssl/bcrypt wrappers to make a public key from a modulus and exponent is trivial but probably out of scope for this PR.

Reply:
Yeah, that's right. I'd be happy to use anything when it becomes available. For now, I'll add an unsupported-ifdef.